modules/gapi/src/compiler/passes/pattern_to_substitute_matching.cpp

```diff
@@ -0,0 +1,176 @@
+        size--;
+    }
+    return matched;
+}
+
+SubgraphMatch::M matchDataNodes2(const Graph& pattern, const Graph& substitute,
+    const std::vector<ade::NodeHandle>& patternNodes,
+    std::vector<ade::NodeHandle> substituteNodes) {
+    SubgraphMatch::M matched;
+    for (auto it : ade::util::zip(patternNodes, substituteNodes)) {
+        const auto& pNode = std::get<0>(it);
+        const auto& sNode = std::get<1>(it);
+        if (!compareDataNodes(pNode, pattern.metadata(pNode), sNode, substitute.metadata(sNode))) {
+            return {};
+        }
+        matched.insert({ pNode, sNode });
+    }
+    return matched;
+}
+
+}  // anonymous namespace
+
+SubgraphMatch findPatternToSubstituteMatch(const Graph& pattern, const Graph& substitute) {
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = pattern.metadata().get<Protocol>().in_nhs;
+    const auto& patternDataOutputs = pattern.metadata().get<Protocol>().out_nhs;
+
+    const auto& substituteDataInputs = substitute.metadata().get<Protocol>().in_nhs;
+    const auto& substituteDataOutputs = substitute.metadata().get<Protocol>().out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
+        || patternDataOutputs.size() != substituteDataOutputs.size()) {
+        return {};
+    }
+
+    // for each pattern input we must find a corresponding substitute input
+    auto matchedDataInputs = matchDataNodes(pattern, substitute, patternDataInputs,
+        substituteDataInputs);
+    // if nothing found, abort
+    if (matchedDataInputs.empty()) {
+        return {};
+    }
+    auto matchedDataOutputs = matchDataNodes(pattern, substitute, patternDataOutputs,
+        substituteDataOutputs);
+    // if nothing found, abort
+    if (matchedDataOutputs.empty()) {
+        return {};
+    }
+
+    //---------------------------------------------------------------
+    // Construct SubgraphMatch object
+    SubgraphMatch match;
+    match.inputDataNodes = std::move(matchedDataInputs);
+    match.outputDataNodes = std::move(matchedDataOutputs);
+    return match;
+}
+
+SubgraphMatch findPatternToSubstituteMatch(const cv::gimpl::GModel::Graph& pattern,
+    const cv::gimpl::GModel::Graph& substitute,
+    const cv::gimpl::Protocol& patternP,
+    const cv::gimpl::Protocol& substituteP) {
+    //---------------------------------------------------------------
+    // Match data nodes which start and end our pattern and substitute
+    const auto& patternDataInputs = patternP.in_nhs;
+    const auto& patternDataOutputs = patternP.out_nhs;
+
+    const auto& substituteDataInputs = substituteP.in_nhs;
+    const auto& substituteDataOutputs = substituteP.out_nhs;
+
+    // if number of data nodes doesn't match, abort
+    if (patternDataInputs.size() != substituteDataInputs.size()
```

Comment:
Looks like you can move the rest of the function below to another function, to prevent copypasting (both version of `findPatternToSubstituteMatch` got that).

Reply:
deleted the 2nd method